### DISCUSSION TITLE
Thirdparty integrations and resources

### DISCUSSION THREAD

**Discussion by mikke89** [mikke89/RmlUi#186] [Show and tell]
Hi all.

I've been collecting some resources and thirdparty integrations for RmlUi. They may be useful for people making their own integrations, or just for having a look around out of curiosity.

**Open source integrations and backends**
- [Aurora](https://github.com/encounter/aurora) (WebGPU, [RenderInterface](https://github.com/encounter/aurora/blob/main/lib/rmlui/WebGPURenderInterface.cpp))
- [TuRml](https://github.com/Tuyuji/TuRml) (RmlUi backend for [O3DE](https://o3de.org/))
- [RmlOgre](https://github.com/nimble0/RmlOgre) (RmlUi render interface for [OGRE-Next](https://github.com/OGRECave/ogre-next))
- [Recoil Engine](https://github.com/beyond-all-reason/RecoilEngine) (Open source RTS engine, [RmlUi integration](https://github.com/beyond-all-reason/RecoilEngine/tree/master/rts/Rml))
- [Godot RmlUi](https://github.com/ashifolfi/Godot-RmlUi) (A GDExtension port of RmlUi to [Godot](https://godotengine.org/) 4.2+)
- [rmlui-bgfx](https://github.com/martonp96/rmlui-bgfx) (RmlUi backend for [bgfx](https://github.com/bkaradzic/bgfx), compatible with D3D11, Vulkan, Metal)
- [DiligentEngineXRmlUI](https://github.com/Daspien27/DiligentEngineXRmlUI). (RmlUi backend for [Diligent engine](https://github.com/DiligentGraphics/DiligentEngine))
- [rbfx](https://github.com/rbfx/rbfx). (Game engine based on [Urho3D](http://urho3d.github.io/). [RmlUi integration](https://github.com/rbfx/rbfx/tree/master/Source/Urho3D/RmlUI))
- [Raylib+RmlUi](https://github.com/NimbusFox/raylib-RmlUi). (RmlUi backend for [raylib](https://www.raylib.com/))
- [QFusion](https://qfusion.github.io/qfusion/). (OpenGL, [RenderInterface](https://github.com/Qfusion/qfusion/blob/master/source/ui/kernel/ui_renderinterface.cpp))
- [RavEngine](https://github.com/RavEngine/RavEngine). (bgfx, [RenderInterface](https://github.com/RavEngine/RavEngine/blob/master/src/RenderEngine_RML.cpp))
- [Rootex](https://github.com/sdslabs/Rootex). (DirectX11, [RenderInterface](https://github.com/sdslabs/Rootex/blob/master/rootex/core/ui/custom_render_interface.cpp))
- [ezEngine](https://github.com/ezEngine/ezEngine). ([RmlUi plugin source](https://github.com/ezEngine/ezEngine/tree/dev/Code/EnginePlugins/RmlUiPlugin/Implementation), [RmlUi documentation and sample](http://ezengine.net/pages/docs/ui/rmlui.html))
- [Kisak-Strike](https://lwss.github.io/Kisak-Strike/). (Unofficial CS:GO port, OpenGL, [Integration](https://github.com/SwagSoftware/Kisak-Strike/tree/master/rocketui))
- [Conquer Space](https://github.com/Conquer-Space/Conquer-Space). (OpenGL, [RenderInterface](https://github.com/Conquer-Space/Conquer-Space/blob/main/src/engine/ui/rmlrenderinterface.cpp))
- [RmlUE4](https://github.com/ZhuRong-HomoStation/RmlUE4). (Unreal Engine 4 integration, Chinese)

**Games and applications**
- [LichtFeld Studio](https://lichtfeld.io) ([github](https://github.com/MrNeRF/LichtFeld-Studio))
- [vkQuake + RmlUi](https://github.com/bradenleague/vkQuake-RmlUi) ([discussion post](https://github.com/mikke89/RmlUi/discussions/895#discussion-9462869)) ([teaser trailer](https://youtu.be/vEAeOuetTQ0))
- [Gearmulator](https://github.com/dsp56300/gearmulator) DSP56300 emulators by [The Usual Suspects](https://dsp56300.wordpress.com/) ([RmlUi integration post](https://dsp56300.wordpress.com/2025/08/19/rmlui-new-gui-rendering-engine-skinning-system/))
- [ROSE Online](https://www.roseonlinegame.com/) ([RmlUi gallery post](https://github.com/mikke89/RmlUi/discussions/184#discussioncomment-15292206)) ([Forum post](https://forum.roseonlinegame.com/topic/7238-patch-notes-2025-09-30-luna-clan-field-updates-ui-improvements-fixes/))
- [WOTInspector](https://armor.wotinspector.com/en/pc/12305-e-50-ausf-m/) ([RmlUi gallery post](https://github.com/mikke89/RmlUi/discussions/184#discussioncomment-15278997)) ([iOS](https://apps.apple.com/app/armor-inspector-wot-pc-blitz/id905734645)) ([Android](https://play.google.com/store/apps/details?id=com.dreamfarmgames.wot.armorinspector.android2))
- [Alchemist](https://docs.fivem.net/docs/alchemist/) by [Cfx.re](https://cfx.re/) (Rockstar Games) ([Forum post](https://forum.cfx.re/t/introducing-alchemist-for-fivem/5366795))
- [Singing Challenge](https://uptochallenge.com/sc/) by @gleblebedev 
- [The Thing: Remastered](https://nightdivestudios.com/the-thing-remastered/) by [Nightdive Studios](https://nightdivestudios.com/)
- [Killing Time: Resurrected](https://nightdivestudios.com/killing-time-resurrected/) by [Nightdive Studios](https://nightdivestudios.com/)
- [Beyon All Reason](https://www.beyondallreason.info/) ([Blog post on RmlUi integration](https://www.beyondallreason.info/microblogs/94))
- [TruckersMP](https://truckersmp.com/) ([Blog post on chat system implemented with RmlUi](https://truckersmp.com/blog/296))
- [Unvanquished](https://unvanquished.net/) ([Blog post on RmlUi integration](https://unvanquished.net/rmlui-moving-onto-a-new-rocket/))
- [alt:V installer](https://altv.mp/)
- [Portal Dogs](https://portaldogs.com/)

**Plugins**
- [RmlSolLua](https://github.com/LoneBoco/RmlSolLua). Unofficial Lua plugin based on sol3, supports Lua 5.1+ and LuaJIT/MoonJIT.

**Tools**
- [RCSS Spritesheet Editor](https://github.com/svenvvv/rcss-sprite-ed).
- [rml-sublime](https://github.com/LoneBoco/rml-sublime). RmlUi syntax definition for Sublime Text 4 ([#304](https://github.com/mikke89/RmlUi/discussions/304)).
- [XML Schema for RML](https://gist.github.com/KaruroChori/0d5fc85c08d8d9528a4294b791cb4e68) ([#760](https://github.com/mikke89/RmlUi/issues/760)).
- [vscode-rmlui](https://github.com/leximayfield/vscode-rmlui).

**Other resources**
- [vcpkg port](https://vcpkg.link/ports/rmlui). See also the [RmlUi port folder](https://github.com/microsoft/vcpkg/tree/master/ports/rmlui).
- [Conan Center recipe](https://conan.io/center/recipes/rmlui) for RmlUi. (See also [#183](https://github.com/mikke89/RmlUi/issues/183) and [C++ documentation](https://mikke89.github.io/RmlUiDoc/pages/cpp_manual/building_with_cmake.html#conan)).
- [xmake](https://xmake.io/#/getting_started). A cross-platform build utility based on Lua. See also [RmlUi package source](https://github.com/xmake-io/xmake-repo/blob/dev/packages/r/rmlui/xmake.lua).
- [Alternate Conan recipe](https://gitlab.com/b110011/conan-rmlui) and [sample project](https://gitlab.com/b110011/rmlui-invaders).
- [Covellite++](http://unicornum.github.io/Covellitepp/index.html). Cross-platform framework, multiple backends, Russian.
- [Multi-instance RmlUi fork](https://github.com/dsp56300/RmlUi/tree/rmlui_multiinstance). Support for loading multiple RmlUi library instances, [see context](https://github.com/mikke89/RmlUi/issues/489#issuecomment-3180808371).
- [RmlUi data binding alternative made in Lua](https://github.com/nimble0/RmlLuaDataBinding).

Feel free to post your own or missing projects below!

**Comment by xland** [comment]
It would be great if integrate [QuickJs](https://bellard.org/quickjs/) into RmlUi,just like Lua.

**Reply by mikke89** [reply]
Yeah, I know some users have integrated Duktape for JS support. There is more discussion on this topic here: #268. If anyone publishes such a plugin let me know and I'll be sure to add it to the list.

**Comment by ShawnCZek** [comment]
Hey @mikke89,

First of all, huge thanks for all the work you have been doing on this library! It is apparent that a lot of work goes into it and that the library is heading in a great direction. There is a bright feature for this project, I believe.

I am one of the game developers on [TruckersMP](https://truckersmp.com/), which is a multiplayer modification for Euro Truck Simulator 2 and American Truck Simulator. The user interface is one of my responsibilities. Throughout the past months, we have been doing extensive research, searching for the best fit for our needs. Even after considering commercial solutions, RmlUi turned out to be the best candidate, particularly because it is open source and is easy to contribute to, unlike other HTML-like and CSS solutions.

Right now, we are switching over from our previous solution, GWEN, to RmlUi. And the first UI document to be reworked is our chat system. We also used this opportunity to completely rewrite the backend, so it is quite a major change for our community.

Here is a showcase of our new chat: https://youtu.be/sspcPSymg54

The changes include stuff like UI scaling, proper support for all character sets, proper word wrapping, a dynamic autocomplete for player mentions and slash commands (a bit of Lua scripting), and so on. It is incredible what one can achieve with CSS and scripting instead of placing some elements around in a designer.

At TruckersMP, we look forward to what's coming next for this library!

**Reply by mikke89** [reply]
Wow, thank you! This is basically the best compliment the library and I could have received :)

I am very excited to follow the development. That chat box looks very sleek! Nice job! I'm glad to have you onboard, and looking forward to pull requests and discussions.

I'd love to put up some screenshots on the gallery as well, if you don't mind.

**Reply by ShawnCZek** [reply]
Thanks for the kind words!

And sure, it would be an honor to be a part of the gallery! I have uploaded [a screenshot from the game](https://drive.google.com/file/d/1caCHQDwGrJWR5ser96pRHsXjiNcdes6M/view?usp=sharing) to my Google Drive so we avoid any compression (from GitHub or any other service) on the way.

**Reply by mikke89** [reply]
Perfect, thanks! I've added it now. Let me know if you want to change something, or have more screenshots in the future :)

**Reply by ShawnCZek** [reply]
Hey @mikke89,

It has been a while since our last update, but the user interface is obviously not our only project. The last time, we looked at improving the chat system, which we later iterated by adding support for IME. This time, we are improving our "Scoreboard", a panel with a player list.

There were a few issues, many of which were discussed in #653. With your help and a few iterations, we created a decent-looking panel that performs really well, even in areas with hundreds of players. Furthermore, we could easily add features like player icons and action buttons. And everything now scales nicely based on the resolution. We think it is a significant improvement compared to the previous version, as seen in [one of our promotional images](https://drive.google.com/file/d/1M9x4mIAbVTHqHzoNgggVSjE_wWG7oTUd/view?usp=drive_link). Or check out [our blog post](https://truckersmp.com/blog/315) for more details.

Anyway, we are still using the 5.1 version but plan to upgrade soon. Let's see how that goes!

By the way, so that you know, this is not an unusual number of players in one area:

![Screenshot of the top of the Player Panel with a highlighted number of players, which is 217](https://github.com/user-attachments/assets/dcae0182-e820-4302-a803-2a382f91b9cd)

**Reply by mikke89** [reply]
Wow, that's a great looking scoreboard, and a massive improvement. Nice job!

I'm very happy to see that it worked out, despite some challenges along the way. It's a nice real-world learning experience for the library too.

If you have a full resolution version of just the new scoreboard, I'd love to put that on the readme if you'd like.

**Reply by ShawnCZek** [reply]
Sure! I made a few changes to exclude players' names for privacy reasons. I have uploaded [a screenshot from the game](https://drive.google.com/file/d/1iQBUdUQ_ybjzrRFq-Ynhbw1eTMlEtEkL/view?usp=sharing) to my Google Drive.

There are a few misalignments of icons, but those should be resolved once we move to 6.0.

**Comment by ashifolfi** [comment]
Hey! I've had my eye on this library for a while and actually just started learning how to use it a few days ago. I ended up creating a Godot addon to add support for RmlUi via GDExtension. It's still quite early on but I might as well share this with others now that it's sort of useful.

[Source Code](https://github.com/ashifolfi/Godot-RmlUi)

![](https://github.com/ashifolfi/Godot-RmlUi/raw/main/.github/screenshots/3ddemo.png)

**Reply by mikke89** [reply]
That's awesome, nice work! Thanks a lot for posting, and for sharing the code :) I have added it to the list. Great to see the interest from the Godot community, and looking forward to following the progress.

**Comment by ArthurVasseur** [comment]
Hey! There is also an integration on xmake https://xmake.io/#/about/introduction

https://github.com/xmake-io/xmake-repo/blob/dev/packages/r/rmlui/xmake.lua

**Reply by mikke89** [reply]
Hey, that's great to see! Didn't know about this one, thanks for the heads-up. I've added it to the list now.

**Comment by badosu** [comment]
@mikke89 First thank you so much for RmlUi, it was integrated on [Recoil RTS Engine](https://github.com/beyond-all-reason/RecoilEngine/) using [RmlSolLua](https://github.com/LoneBoco/RmlSolLua) and by all indications it should make the lives of gamedevs using Recoil much easier when writing gui components.

In hopes of finding a way to provide lua annotations for [lua-language-server](https://luals.github.io/wiki/annotations/) and our rmlui lua-based integration I searched this repos issues, pull requests and now discussions. Since this is related to resources, and maybe I just was not able to find them, I feel compelled to ask if there's some resource (to be added here) for lua documentation in a format readable by lua-language-server.

If there's none, I was unable to find how your liquid generator data [is curated](https://github.com/mikke89/RmlUiDoc/blob/master/pages/lua_manual/rmlui_lua_documentation.json), is it manually edited? I was hoping to get a clue on how to achieve the type library myself using that pipeline. I'd reckon would be a valuable resource, thank you!

**Reply by mikke89** [reply]
Very cool, thanks for letting me know about this project! I added it to the list above. Is it the engine behind Beyond All Reason?

As for the Lua documentation. I'm not really familiar with the format used by lua-language-server. And you're right, at this point the API reference is curated manually. The markdown document is the true source for this info.

Originally, the Lua reference was coupled together with some reg.ex. and Python magic from the Lua plugin source, and also using a combination of C++ documentation and the old Python API documentation (no longer used) for descriptions. It was a bit of a mess, and did require some manual curation, so eventually I didn't end up making something automatic. The first draft was generated from [this Python script](https://github.com/mikke89/RmlUiDoc/blob/master/pages/lua_manual/format_docs.py) and the JSON you linked. But later on, this has been manually edited. 

The Lua plugin source code is actually relatively structured, so with some extra tagging and descriptions and such, I don't think it would be too complicated to automate it all. Maybe we could have done that originally, but it hasn't really been too much work to keep it up-to-date manually, so I don't think it would have payed off at this point. Anyway, that was Today's quick history lesson :)

I'd be very interested to see if somebody can make something out of the language server annotations, let me know if I can help in any way.

**Reply by badosu** [reply]
Thank you for clarifying!

> Very cool, thanks for letting me know about this project! I added it to the list above. Is it the engine behind Beyond All Reason?

Yep! It powers Zero-K and some other games as well. Currently BAR is the only game with some expectation of getting RmlUI components soon, at least one gamedev starting their own game has picked interest on RmlUI as a viable option from scratch (instead of raw Lua GL calls or some Lua framework on top).

> I'd be very interested to see if somebody can make something out of the language server annotations, let me know if I can help in any way.

In Recoil, we annotate the Lua API directly into the cpp source code (to make docs parity easier to maintain), [example](https://github.com/beyond-all-reason/RecoilEngine/commit/2be21b203cf7bbc4b590e22151dfda4773b26cad). Then using [lua-doc-extractor](https://github.com/rhys-vdw/lua-doc-extractor) (the author made it for this purpose, previously we used LDoc, but it only provided html doc generation which is not very flexible for our needs) we extract the relevant Lua API Documentation from the source and glue it with some other pieces on our workflow.

The end result is autogenerated from source [docs](https://beyond-all-reason.github.io/RecoilEngine/lua-api) (the templating is still a bit rough as we're transitioning to the new format) and [lua-library](https://github.com/beyond-all-reason/recoil-lua-library). The latter has been a major improvement for gamedevs (in BAR or other consumers of Recoil) so having our exposed `RmlUi` global documented as well would be a major part of this effort.

Thanks for the pointers, we have [an issue](https://github.com/beyond-all-reason/RecoilEngine/issues/1672) related to this, perhaps we'll try and use the json file to generate a valid lua types library for RmlUi, will post here if/when that happens.

**Reply by rhys-vdw** [reply]
Hi, I'm the author of lua-doc-extractor. You may find that it's useful to your project. It's a little immature at present but it does have a massive stress test with RecoilEngine (and I am adding features ongoing). If there is any interest in integrating it into this project let me know.

**Reply by mikke89** [reply]
Hey, that's very cool! Thanks for letting me know.

I'd be very interested to use something like this. If we could backport the documentation to the source, and have that generate a close equivalent to what we have, that would be super cool. Right now I think it would take more time and effort than I can dedicate to this, but I'd certainly be willing to assist if somebody wants to have a go at that.

I also see that you are adding docstrings to your RmlLuaSol fork: https://github.com/beyond-all-reason/RecoilEngine/pull/2247
Maybe it would make sense for us to switch our Lua plugin to this at some stage?

**Reply by rhys-vdw** [reply]
Yeah, I think it would make sense to upstream those docs once @SlashScreen is done with them.

The advantage of using this system is that it generates [definition files](https://luals.github.io/wiki/definition-files/) for lua-language-server which gives intellisense in the editor. You can see the output for recoil here: https://github.com/beyond-all-reason/recoil-lua-library/

You'd have to make a separate repo like this too, and then it can be used by a submodule by your API consumers.

I probably don't have the bandwidth to actually implement the system for you, but it might be a case of copying the relevant workflows from the RecoilEngine repo and adapting them.

**Reply by mikke89** [reply]
I see, that sounds very cool. I am definitely interested in seeing this upstreamed.

**Reply by badosu** [reply]
> I also see that you are adding docstrings to your RmlLuaSol fork: [beyond-all-reason/RecoilEngine#2247](https://github.com/beyond-all-reason/RecoilEngine/pull/2247) Maybe it would make sense for us to switch our Lua plugin to this at some stage?

Can't say for the effort related to using `sol2` in RmlUi source, it's a non-trivial effort but I can see benefits (easier to reason about and maintain) and drawbacks (tied to whatever `sol2` supports, e.g. [this issue](https://github.com/ThePhD/sol2/issues/1694)).

In any manner: inline docs using `lua-doc-extractor` for pulling Lua documentation from source (and populating a lua library for [lua-language-server](https://github.com/LuaLS/lua-language-server) or [EmmyLua-LanguageServer](https://github.com/EmmyLua/EmmyLua-LanguageServer), with automatic html/markdown docgen for site) is orthogonal to Lua C API wrapper choice (at least while `sol2` [does not support automatic docgen](https://github.com/ThePhD/sol2/issues/1537), which it won't anytime soon).

One can already document using normal C Lua API [as we do on Recoil](https://github.com/beyond-all-reason/RecoilEngine/blob/master/rts/Lua/LuaSyncedRead.cpp#L1378-L1386), or `sol2` [as we also do on Recoil](https://github.com/beyond-all-reason/RecoilEngine/blob/master/rts/Lua/LuaVBOImpl.cpp#L523-L534).

> I'd be very interested to use something like this. If we could backport the documentation to the source, and have that generate a close equivalent to what we have, that would be super cool.

Example of an action item: **document `Element:GetElementsByName`** _(pardon me if my understanding of the actual RmlUi API is inaccurate, but the point still stands, `Element[]` below could be also `table<integer, Element> elements A 0-indexed list of matching elements`)_

1. Right above [this line](https://github.com/mikke89/RmlUi/blob/master/Source/Lua/Element.cpp#L176), (where this function is defined, could be some other source if it makes more sense, up to the author), add doc as this (assuming `@class Element` exists somewhere):
```cpp
/*** Returns a list of all descendant elements matched by name.
 *
 * @function Element:GetElementsByName
 * @param tag_name string The tag name to be matched against
 * @return Element[] elements The list of matching elements
 */
```
2. Have github workflow actions (or a manual process) to extract these docs with `lua-doc-extractor` and populate the RmlUi "lua library" (types for language server) and/or generate html/markdown pages from them.

> I see, that sounds very cool. I am definitely interested in seeing this upstreamed.

Glad to read that, thank you! Recoil requires the whole exposed API to be documented anyway, so at some point either we'll have our own `RmlUi` base documentation (that can be in some way upstreamed) or a strategy to do this in `RmlUi` itself (as PRs to this repo) and pulled to our docgen build process.

**Reply by espkk** [reply]
> tied to whatever `sol2` support

I don't think it's `sol2` that doesn't support `pairs`. It is lua 5.1 that doesn't support `__pairs` (luajit supports them with `DLUAJIT_ENABLE_LUA52COMPAT` flag, but has slightly different behaviour iirc)

By the way, I'm also interested in seeing this upstreamed/in rmlui (and helping if needed)

**Comment by Tellusim** [comment]
Hello,

Thank you for the great library!

RmlUi plugin is now available in Tellusim SDK: https://github.com/Tellusim/Tellusim_Core_SDK/tree/main/plugins/interface/rmlui
It is compatible with Vulkan, D3D12, D3D11, Metal, OpenGL, OpenGLES, and WebGPU backends and supports all rendering features.

https://tellusim.com/demos/webgpu-20260728/interface_rmlui/
